### PR TITLE
Swap slot

### PR DESCRIPTION
**Pull request by gavinfish**
- Add swapping slots feature
- Fix some legacy integration tests

### CODE REVIEW

File: src/main/java/com/microsoft/jenkins/appservice/WebAppDeploymentRecorder.java [hunk 1]
```diff
@@ -360,6 +359,9 @@ public FormValidation doCheckFilePath(
                 @QueryParameter String azureCredentialsId,
                 @QueryParameter String resourceGroup,
                 @QueryParameter String appName) {
+            if (StringUtils.isBlank(azureCredentialsId)) {
+                return FormValidation.ok();
```
Comment (yungezz) [inline src/main/java/com/microsoft/jenkins/appservice/WebAppDeploymentRecorder.java]: isblank is ok?

File: src/main/java/com/microsoft/jenkins/appservice/WebAppSwapSlotsRecorder.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+    public String getAppName() {
+        return appName;
+    }
+
+    public String getSourceSlotName() {
+        return sourceSlotName;
+    }
+
+    public String getTargetSlotName() {
```
Comment (yungezz) [inline src/main/java/com/microsoft/jenkins/appservice/WebAppSwapSlotsRecorder.java]: preserve virtual network?

File: src/main/java/com/microsoft/jenkins/appservice/WebAppSwapSlotsRecorder.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+            throw new AbortException(e.getMessage());
+        }
+
+        context.executeCommands();
+
+        if (!context.getLastCommandState().isError()) {
+            listener.getLogger().println(String.format("Done swapping Azure Web "
+                    + "App deployment slots between %s and %s.", sourceSlotName, targetSlotName));
+        } else {
+            throw new AbortException("Azure Web App swapping slots failed.");
```
Comment (yungezz) [inline src/main/java/com/microsoft/jenkins/appservice/WebAppSwapSlotsRecorder.java]: add exception detail?

**Comment by gavinfish** [comment]
> there's auto-swap function, and preserve_vnet function during swap.

Yes, auto-swap function and swap with preview function are only available for windows web app now. No preserve_vnet option in portal.

File: src/main/java/com/microsoft/jenkins/appservice/WebAppDeploymentRecorder.java (hunk 1 above)
Comment (gavinfish) [reply src/main/java/com/microsoft/jenkins/appservice/WebAppDeploymentRecorder.java]: When a user creates this command in the UI for the first time, the credentialId will be blank. Just ignore this parameter check for this case. The implementation here need credentailId to fetch web app information.

File: src/main/java/com/microsoft/jenkins/appservice/WebAppSwapSlotsRecorder.java (hunk 3 above)
Comment (gavinfish) [reply src/main/java/com/microsoft/jenkins/appservice/WebAppSwapSlotsRecorder.java]: The exception detail has already been recorded in the executed command. In the current design, no error message is available at this context.

File: src/main/java/com/microsoft/jenkins/appservice/WebAppSwapSlotsRecorder.java (hunk 2 above)
Comment (gavinfish) [reply src/main/java/com/microsoft/jenkins/appservice/WebAppSwapSlotsRecorder.java]: There is no direct method or parameters can controller this as the REST api does. Have create an issue for the sdk team at Azure/azure-libraries-for-java#742 .

**Pull request merged**